python/src/trezorlib/cli/firmware.py

```diff
@@ -166,39 +281,212 @@ def version_str(version):
+        click.echo("Update failed: {}".format(e))
+        sys.exit(3)
 
-    return url, fingerprint
 
+@click.group(name="firmware")
+def cli():
+    """Firmware commands."""
 
-@click.command()
+
+@cli.command()
 # fmt: off
-@click.option("-f", "--filename", type=click.File("rb"))
-@click.option("-u", "--url")
-@click.option("-v", "--version")
+@click.argument("filename", type=click.File("rb"))
+@click.option("-c", "--check-device", is_flag=True, help="Validate the device compatibility")
+@click.option("--fingerprint", help="Expected firmware fingerprint in hex")
+@click.pass_obj
+# fmt: on
+def verify(
+    obj: TrezorConnection,
+    filename: BinaryIO,
+    check_device: bool,
+    fingerprint: str,
+) -> None:
+    """Verify the integrity of the firmware data stored in a file.
+
+    By default the device is not checked and does not need to be connected.
+    Its validation must be specified.
+
+    In case of validation failure exits with the appropriate exit code.
+    """
+    # Deciding if to take the device into account
+    if check_device:
+        with obj.client_context() as client:
+            bootloader_onev2 = _is_bootloader_onev2(client)
+            trezor_major_version = client.features.major_version
+    else:
+        bootloader_onev2 = None
+        trezor_major_version = None
+
+    firmware_data = filename.read()
+    validate_firmware(
+        firmware_data=firmware_data,
+        fingerprint=fingerprint,
+        bootloader_onev2=bootloader_onev2,
+        trezor_major_version=trezor_major_version,
+    )
+
+
+@cli.command()
+# fmt: off
+@click.option("-o", "--output", type=click.File("wb"), help="Output file to save firmware data to")
+@click.option("-v", "--version", help="Which version to download")
+@click.option("-s", "--skip-check", is_flag=True, help="Do not validate firmware integrity")
+@click.option("--beta", is_flag=True, help="Use firmware from BETA channel")
+@click.option("--bitcoin-only", is_flag=True, help="Use bitcoin-only firmware (if possible)")
+@click.option("--fingerprint", help="Expected firmware fingerprint in hex")
+# fmt: on
+@with_client
+def download(
+    client: TrezorClient,
+    output: BinaryIO,
+    version: str,
+    skip_check: bool,
+    fingerprint: str,
+    beta: bool,
+    bitcoin_only: bool,
+) -> None:
+    """Download and save the firmware image.
+
+    Validation is done by default, can be omitted by "-s" or "--skip-check".
+    """
+    url, fp = find_best_firmware_version(
+        client=client, version=version, beta=beta, bitcoin_only=bitcoin_only
+    )
+
+    firmware_data = download_firmware_data(url)
+
+    # When fingerprint and output file were not supplied, take them from SL servers
+    if not fingerprint:
+        fingerprint = fp
+    if not output:
+        output = open(_get_file_name_from_url(url), "wb")
+
+    output.write(firmware_data)
+    output.close()
+    click.echo("Firmware saved under {}.".format(output.name))
+
+    if not skip_check:
```

Comment:
validate first, write file only if validation succeeds

Reply:
Modified in 2ae4